File: code-monkey/src/tools/ca_imports_tool.py
import json
from typing import Dict, Any
from .ca_tool_base import CATool
from .ast_parser import ASTParser
from .utils import resolve_file_path, resolve_module_path, get_module_name

class CAImportsTool(CATool):
    name = "ca_analyze_imports"
    description = "Analyze the imports in Python files"
    input_schema = {
        "type": "object",
        "properties": {
            "files": {
                "type": "array",
                "items": {"type": "string"},
                "description": "List of relative file paths to analyze",
            },
        },
        "anyOf": [{"required": ["files"]}, {"required": ["modules"]}],
    }

    def __init__(self, parser: ASTParser):
        super().__init__()
        self.parser = parser

    def handle_tool_call(self, input: Dict[str, Any]) -> str:
        files = input.get("files", [])
        modules = input.get("modules", [])

        all_files = [resolve_file_path(f) for f in files] + [
            resolve_module_path(m) for m in modules
        ]

        imports_analysis = {}
        for file_path in all_files:
            tree = self.parser.parse_file(file_path)
            module_name = get_module_name(file_path)
            imports_analysis[module_name] = self.parser.get_imports(tree)

        return json.dumps(imports_analysis, indent=1)

File: code-monkey-ts/tools/ca/ca_imports_tool.ts
import { z } from 'zod';
import { AsyncCallbackManagerForToolRun } from '@langchain/core/callbacks/manager';
import { tool } from '@langchain/core/tools';
import { CATool } from './ca_tool';
import { resolveFilePath, resolveModulePath, getModuleName } from '../deps';
import { instrument } from '../../instrumentation';

const CAImportsToolInputSchema = z.object({
    files: z.array(z.string()).nullable().describe("List of relative file paths to analyze"),
    modules: z.array(z.string()).nullable().describe("List of modules to analyze")
}).refine(data => data.files != null || data.modules != null, {
    message: "Either 'files' or 'modules' must be provided"
});

type CAImportsToolInput = z.infer<typeof CAImportsToolInputSchema>;

export class CAImportsTool extends CATool {
    name = "ca_analyze_imports";
    description = "Analyze the imports in Python files";

    constructor() {
        super();
    }

    @instrument("Tool._run", ["files", "modules"], { attributes: { tool: "CAImportsTool" } })
    async _run(
        input: CAImportsToolInput,
        runManager?: AsyncCallbackManagerForToolRun
    ): Promise<string> {
        const { files = [], modules = [] } = input;

        const allFiles = [
            ...files.map(resolveFilePath),
            ...modules.map(resolveModulePath)
        ];

        const importsAnalysis: Record<string, unknown> = {};
        for (const filePath of allFiles) {
            const moduleName = getModuleName(filePath);
            importsAnalysis[moduleName] = await this.parser.getImports(filePath);
        }

        return JSON.stringify(importsAnalysis, null, 1);
    }
}

export const caImportsTool = tool({
    name: "ca_analyze_imports",
    description: "Analyze the imports in Python files",
    schema: CAImportsToolInputSchema,
    func: async (input: CAImportsToolInput, runManager?: AsyncCallbackManagerForToolRun) => {
        const tool = new CAImportsTool();
        return tool._run(input, runManager);
    }
});
